### PR TITLE
add cmake flag for testing, add common definitions file

### PR DESCRIPTION
**Pull request by pkarasev3**
- added some cmake helper functionality . 
- added BUILD_TESTS flag to configure from gui .
- added ctest hint so one can see the nicely done matrix of tests available.

### CODE REVIEW

File: cmake/common.cmake [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+set(Green    "${Esc}[92m"     CACHE STRING "green color"    FORCE)
+set(Blue     "${Esc}[1;94m"   CACHE STRING "Blue color"     FORCE)
+set(BoldRed  "${Esc}[1;5;91m" CACHE STRING "bold-red color" FORCE)
+set(NoColor  "${Esc}[0m"      CACHE STRING "reset color"    FORCE)
+
+if(NOT WIN32)
+
+    execute_process(COMMAND   ${CMAKE_C_COMPILER}  -dumpversion   OUTPUT_VARIABLE   GCC_VERSION)
+    if(GCC_VERSION VERSION_GREATER 4.9 OR GCC_VERSION VERSION_EQUAL 4.9)
+        set(STDLIBC  "-std=c++1y")
```
Comment (stevengj) [inline cmake/common.cmake]: Why do we need this?  C++ 2003 is probably fine.... we don't use any recent C++ features, so there should be no need to specify `-std=c++XX`.

File: cmake/common.cmake [hunk 2]
```diff
@@ -0,0 +1,35 @@
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib")
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib")
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/bin")
+    
+string(ASCII 27 Esc)
+set(Green    "${Esc}[92m"     CACHE STRING "green color"    FORCE)
+set(Blue     "${Esc}[1;94m"   CACHE STRING "Blue color"     FORCE)
```
Comment (stevengj) [inline cmake/common.cmake]: Do these terminal codes work on Windows?

File: cmake/common.cmake (hunk 1 above)
Comment (pkarasev3) [inline cmake/common.cmake]: "we don't use ..."  presumably that could change right? :-)  

 It's nice to be able to use bounds-checking structures with simple syntax. 
 For example: 
       `**std::vector<double> coefficients = {1.0, 2.0,  rand()*1.0 };** ` 
  looks a lot simpler than however you'd do it in C++03 (for-loop?), and safer than a raw C-style array. 

Also, on Windows, visual studio 2012 by default will enable C++11, while visual studio 2013 enables C++14. These gcc flags are to align the default compiler behavior across platforms.

File: cmake/common.cmake [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+set(BoldRed  "${Esc}[1;5;91m" CACHE STRING "bold-red color" FORCE)
+set(NoColor  "${Esc}[0m"      CACHE STRING "reset color"    FORCE)
+
+if(NOT WIN32)
+
+    execute_process(COMMAND   ${CMAKE_C_COMPILER}  -dumpversion   OUTPUT_VARIABLE   GCC_VERSION)
```
Comment (jschueller) [inline cmake/common.cmake]: This is not really portable "-dumpversion" is only available with GCC (maybe clang).

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,13 @@
 add_custom_target (tests)
 
+if(BUILD_TESTS)
+    set(TESTING_FLAG  "")
+else()
+    set(TESTING_FLAG EXCLUDE_FROM_ALL)
+endif()
+
```
Comment (jschueller) [inline test/CMakeLists.txt]: wouldn't be easier to exlude the compiled tests in all cases and provide a custom target ?

File: cmake/common.cmake (hunk 1 above)
Comment (stevengj) [inline cmake/common.cmake]: We use C++ so little in NLopt that the benefits of using features from the latest C++ standard don't seem to outweigh the cost of dropping support for usage by old programs that require earlier versions of the standard.

**Pull request closed without merging**